File: packages/ink-graves/src/cell.ts
/**
 * The Graves handwriting cell: 3 stacked LSTMs + Gaussian window attention
 * + mixture density head. A line-by-line port of the MLX reference
 * (graves_handwriting_mlx/model.py and modules.py); concat row orders must
 * match the saved kernel layouts exactly:
 *
 *   lstm1 rows:     [w_prev(73), x(3), h1_prev(400)]          -> (476, 1600)
 *   attention rows: [w_prev(73), x(3), h1(400)]               -> (476, 30)
 *   lstm2 rows:     [x(3), h1(400), w(73), h2_prev(400)]      -> (876, 1600)
 *   lstm3 rows:     [x(3), h2(400), w(73), h3_prev(400)]      -> (876, 1600)
 *   gmm rows:       [h3(400)]                                 -> (400, 121)
 *
 * Everything is batch-size 1 and allocation-free per step.
 */

import type { ModelAssets } from "./weights.js";
import type { Rng } from "./rng.js";

export const HIDDEN = 400;
export const ATTENTION_MIXTURES = 10;
export const OUTPUT_MIXTURES = 20;
export const ALPHABET_SIZE = 73;
export const MAX_CHARS = 120;

const GATES = 4 * HIDDEN;
const EPSILON = 1e-8;
const SIGMA_FLOOR = 1e-4;

export interface CellState {
  h1: Float32Array;
  c1: Float32Array;
  h2: Float32Array;
  c2: Float32Array;
  h3: Float32Array;
  c3: Float32Array;
  kappa: Float32Array;
  w: Float32Array;
  phi: Float32Array;
}

export interface MdnParams {
  pi: Float32Array;
  muX: Float32Array;
  muY: Float32Array;
  sigmaX: Float32Array;
  sigmaY: Float32Array;
  rho: Float32Array;
  eos: number;
}

function sigmoid(v: number): number {
  return 1 / (1 + Math.exp(-v));
}

function softplus(v: number): number {
  return v > 30 ? v : Math.log1p(Math.exp(v));
}

/** y += x @ kernel[rowOffset : rowOffset + xLength], kernel row-major (rows, nOut). */
function accumulate(
  y: Float32Array,
  kernel: Float32Array,
  rowOffset: number,
  nOut: number,
  x: Float32Array,
  xLength: number,
): void {
  for (let i = 0; i < xLength; i++) {
    const xi = x[i];
    if (xi === 0) continue;
    const base = (rowOffset + i) * nOut;
    for (let j = 0; j < nOut; j++) y[j] += kernel[base + j] * xi;
  }
}

/** In-place LSTM update from a fused gate buffer, TF gate order (i, j, f, o). */
function applyLstm(gates: Float32Array, hidden: Float32Array, cell: Float32Array): void {
  for (let m = 0; m < HIDDEN; m++) {
    const inputGate = sigmoid(gates[m]);
    const candidate = Math.tanh(gates[HIDDEN + m]);
    const forgetGate = sigmoid(gates[2 * HIDDEN + m]);
    const outputGate = sigmoid(gates[3 * HIDDEN + m]);
    const newCell = forgetGate * cell[m] + inputGate * candidate;
    cell[m] = newCell;
    hidden[m] = outputGate * Math.tanh(newCell);
  }
}

export class Cell {
  private readonly k1: Float32Array;
  private readonly b1: Float32Array;
  private readonly k2: Float32Array;
  private readonly b2: Float32Array;
  private readonly k3: Float32Array;
  private readonly b3: Float32Array;
  private readonly kAtt: Float32Array;
  private readonly bAtt: Float32Array;
  private readonly kGmm: Float32Array;
  private readonly bGmm: Float32Array;

  // Per-step scratch, reused across calls.
  private readonly gates = new Float32Array(GATES);
  private readonly attRaw = new Float32Array(3 * ATTENTION_MIXTURES);
  private readonly gmmRaw = new Float32Array(6 * OUTPUT_MIXTURES + 1);
  private readonly windowScratch = new Float32Array(ALPHABET_SIZE);
  private readonly inputScratch = new Float32Array(3);

  constructor(assets: ModelAssets) {
    const tensor = (name: string): Float32Array => {
      const found = assets.tensors.get(name);
      if (!found) throw new Error(`missing tensor ${name}`);
      return found.data;
    };
    this.k1 = tensor("lstm1_kernel");
    this.b1 = tensor("lstm1_bias");
    this.k2 = tensor("lstm2_kernel");
    this.b2 = tensor("lstm2_bias");
    this.k3 = tensor("lstm3_kernel");
    this.b3 = tensor("lstm3_bias");
    this.kAtt = tensor("attention_weights");
    this.bAtt = tensor("attention_biases");
    this.kGmm = tensor("gmm_weights");
    this.bGmm = tensor("gmm_biases");
  }

  initialState(): CellState {
    return {
      h1: new Float32Array(HIDDEN),
      c1: new Float32Array(HIDDEN),
      h2: new Float32Array(HIDDEN),
      c2: new Float32Array(HIDDEN),
      h3: new Float32Array(HIDDEN),
      c3: new Float32Array(HIDDEN),
      kappa: new Float32Array(ATTENTION_MIXTURES),
      w: new Float32Array(ALPHABET_SIZE),
      phi: new Float32Array(MAX_CHARS),
    };
  }

  newMdnParams(): MdnParams {
    return {
      pi: new Float32Array(OUTPUT_MIXTURES),
      muX: new Float32Array(OUTPUT_MIXTURES),
      muY: new Float32Array(OUTPUT_MIXTURES),
      sigmaX: new Float32Array(OUTPUT_MIXTURES),
      sigmaY: new Float32Array(OUTPUT_MIXTURES),
      rho: new Float32Array(OUTPUT_MIXTURES),
      eos: 0,
    };
  }

  /**
   * One timestep. Mutates `state` in place. `chars` holds alphabet indices
   * (the encoded text, zero-padded); `charLength` is the true encoded length.
   */
  step(state: CellState, dx: number, dy: number, eos: number, chars: Int32Array, charLength: number): void {
    const x = this.inputScratch;
    x[0] = dx;
    x[1] = dy;
    x[2] = eos;
    const { gates, attRaw } = this;

    // LSTM 1: [w_prev, x, h1_prev]
    gates.set(this.b1);
    accumulate(gates, this.k1, 0, GATES, state.w, ALPHABET_SIZE);
    accumulate(gates, this.k1, ALPHABET_SIZE, GATES, x, 3);
    accumulate(gates, this.k1, ALPHABET_SIZE + 3, GATES, state.h1, HIDDEN);
    applyLstm(gates, state.h1, state.c1);

    // Attention: [w_prev, x, h1] -> softplus -> (alpha, beta, kappa step)
    attRaw.set(this.bAtt);
    accumulate(attRaw, this.kAtt, 0, 30, state.w, ALPHABET_SIZE);
    accumulate(attRaw, this.kAtt, ALPHABET_SIZE, 30, x, 3);
    accumulate(attRaw, this.kAtt, ALPHABET_SIZE + 3, 30, state.h1, HIDDEN);
    for (let i = 0; i < 30; i++) attRaw[i] = softplus(attRaw[i]);
    for (let k = 0; k < ATTENTION_MIXTURES; k++) {
      state.kappa[k] += attRaw[20 + k] / 25.0;
      if (attRaw[10 + k] < 0.01) attRaw[10 + k] = 0.01; // beta floor
    }
    for (let u = 0; u < MAX_CHARS; u++) {
      let sum = 0;
      for (let k = 0; k < ATTENTION_MIXTURES; k++) {
        const diff = state.kappa[k] - u;
        sum += attRaw[k] * Math.exp(-(diff * diff) / attRaw[10 + k]);
      }
      state.phi[u] = sum;
    }
    this.windowScratch.fill(0);
    for (let u = 0; u < charLength; u++) {
      this.windowScratch[chars[u]] += state.phi[u];
    }
    state.w.set(this.windowScratch);

    // LSTM 2: [x, h1, w, h2_prev]
    gates.set(this.b2);
    accumulate(gates, this.k2, 0, GATES, x, 3);
    accumulate(gates, this.k2, 3, GATES, state.h1, HIDDEN);
    accumulate(gates, this.k2, 3 + HIDDEN, GATES, state.w, ALPHABET_SIZE);
    accumulate(gates, this.k2, 3 + HIDDEN + ALPHABET_SIZE, GATES, state.h2, HIDDEN);
    applyLstm(gates, state.h2, state.c2);

    // LSTM 3: [x, h2, w, h3_prev]
    gates.set(this.b3);
    accumulate(gates, this.k3, 0, GATES, x, 3);
    accumulate(gates, this.k3, 3, GATES, state.h2, HIDDEN);
    accumulate(gates, this.k3, 3 + HIDDEN, GATES, state.w, ALPHABET_SIZE);
    accumulate(gates, this.k3, 3 + HIDDEN + ALPHABET_SIZE, GATES, state.h3, HIDDEN);
    applyLstm(gates, state.h3, state.c3);
  }

  /**
   * MDN head with the Graves bias (sharpness) trick. Matches the reference:
   * pi logits scaled by (1 + bias), log-sigmas shifted down by bias, then
   * pi and eos snapped to zero below 0.01 (pi is NOT renormalized).
   */
  mdnParse(h3: Float32Array, bias: number, out: MdnParams): MdnParams {
    const raw = this.gmmRaw;
    raw.set(this.bGmm);
    accumulate(raw, this.kGmm, 0, raw.length, h3, HIDDEN);

    const M = OUTPUT_MIXTURES;
    let maxLogit = -Infinity;
    for (let m = 0; m < M; m++) {
      const logit = raw[m] * (1 + bias);
      out.pi[m] = logit;
      if (logit > maxLogit) maxLogit = logit;
    }
    let total = 0;
    for (let m = 0; m < M; m++) {
      const value = Math.exp(out.pi[m] - maxLogit);
      out.pi[m] = value;
      total += value;
    }
    for (let m = 0; m < M; m++) {
      const p = out.pi[m] / total;
      out.pi[m] = p < 0.01 ? 0 : p;
      out.sigmaX[m] = Math.max(Math.exp(raw[M + m] - bias), SIGMA_FLOOR);
      out.sigmaY[m] = Math.max(Math.exp(raw[2 * M + m] - bias), SIGMA_FLOOR);
      const tanhRho = Math.tanh(raw[3 * M + m]);
      out.rho[m] = Math.min(Math.max(tanhRho, EPSILON - 1), 1 - EPSILON);
      out.muX[m] = raw[4 * M + m];
      out.muY[m] = raw[5 * M + m];
    }
    const eosProb = Math.min(Math.max(sigmoid(raw[6 * M]), EPSILON), 1 - EPSILON);
    out.eos = eosProb < 0.01 ? 0 : eosProb;
    return out;
  }

  /** Draw one (Δx, Δy, eos) from parsed MDN params. */
  mdnSample(params: MdnParams, rng: Rng): [number, number, number] {
    const component = rng.categorical(params.pi);
    const z1 = rng.normal();
    const z2 = rng.normal();
    const rho = params.rho[component];
    const dx = params.muX[component] + params.sigmaX[component] * z1;
    const dy =
      params.muY[component] +
      params.sigmaY[component] * (rho * z1 + Math.sqrt(Math.max(1 - rho * rho, 0)) * z2);
    const eos = rng.uniform() < params.eos ? 1 : 0;
    return [dx, dy, eos];
  }
}

File: packages/ink-swift/Sources/InkGraves/Cell.swift
/// The Graves handwriting cell: 3 stacked LSTMs + Gaussian window attention
/// + mixture density head. A line-by-line port of the TS engine
/// (packages/ink-graves/src/cell.ts), which is itself a port of the MLX
/// reference; concat row orders must match the saved kernel layouts exactly:
///
///   lstm1 rows:     [w_prev(73), x(3), h1_prev(400)]          -> (476, 1600)
///   attention rows: [w_prev(73), x(3), h1(400)]               -> (476, 30)
///   lstm2 rows:     [x(3), h1(400), w(73), h2_prev(400)]      -> (876, 1600)
///   lstm3 rows:     [x(3), h2(400), w(73), h3_prev(400)]      -> (876, 1600)
///   gmm rows:       [h3(400)]                                 -> (400, 121)
///
/// Everything is batch-size 1 and allocation-free per step. The matrix-vector
/// products go through vDSP where available (fast even in debug builds);
/// the nonlinear tails run in Double like the JS engine's number math.

import Foundation
#if canImport(Accelerate)
import Accelerate
#endif

let HIDDEN = 400
let ATTENTION_MIXTURES = 10
let OUTPUT_MIXTURES = 20
let ALPHABET_SIZE = 73
let MAX_CHARS = 120

private let GATES = 4 * HIDDEN
private let EPSILON = 1e-8
private let SIGMA_FLOOR = 1e-4

public final class CellState {
    public var h1 = [Float](repeating: 0, count: HIDDEN)
    public var c1 = [Float](repeating: 0, count: HIDDEN)
    public var h2 = [Float](repeating: 0, count: HIDDEN)
    public var c2 = [Float](repeating: 0, count: HIDDEN)
    public var h3 = [Float](repeating: 0, count: HIDDEN)
    public var c3 = [Float](repeating: 0, count: HIDDEN)
    public var kappa = [Float](repeating: 0, count: ATTENTION_MIXTURES)
    public var w = [Float](repeating: 0, count: ALPHABET_SIZE)
    public var phi = [Float](repeating: 0, count: MAX_CHARS)
}

public final class MdnParams {
    public var pi = [Float](repeating: 0, count: OUTPUT_MIXTURES)
    public var muX = [Float](repeating: 0, count: OUTPUT_MIXTURES)
    public var muY = [Float](repeating: 0, count: OUTPUT_MIXTURES)
    public var sigmaX = [Float](repeating: 0, count: OUTPUT_MIXTURES)
    public var sigmaY = [Float](repeating: 0, count: OUTPUT_MIXTURES)
    public var rho = [Float](repeating: 0, count: OUTPUT_MIXTURES)
    public var eos: Float = 0
}

/// One raw model output row: pen movement delta and end-of-stroke flag.
public struct StrokeOffset: Equatable, Sendable {
    public let dx: Double
    public let dy: Double
    public let eos: Bool

    public init(dx: Double, dy: Double, eos: Bool) {
        self.dx = dx
        self.dy = dy
        self.eos = eos
    }
}

@inline(__always) private func sigmoid(_ v: Double) -> Double {
    1 / (1 + exp(-v))
}

@inline(__always) private func softplus(_ v: Double) -> Double {
    v > 30 ? v : log1p(exp(v))
}

/// y[0..<nOut] += x @ kernel[rowOffset ..< rowOffset + x.count], kernel
/// row-major (rows, nOut).
private func accumulate(
    _ y: inout [Float],
    kernel: [Float],
    rowOffset: Int,
    nOut: Int,
    x: [Float],
    scratch: inout [Float]
) {
    #if canImport(Accelerate)
    kernel.withUnsafeBufferPointer { kernelBuffer in
        x.withUnsafeBufferPointer { xBuffer in
            scratch.withUnsafeMutableBufferPointer { scratchBuffer in
                vDSP_mmul(
                    xBuffer.baseAddress!, 1,
                    kernelBuffer.baseAddress! + rowOffset * nOut, 1,
                    scratchBuffer.baseAddress!, 1,
                    1, vDSP_Length(nOut), vDSP_Length(x.count)
                )
                y.withUnsafeMutableBufferPointer { yBuffer in
                    vDSP_vadd(
                        yBuffer.baseAddress!, 1,
                        scratchBuffer.baseAddress!, 1,
                        yBuffer.baseAddress!, 1,
                        vDSP_Length(nOut)
                    )
                }
            }
        }
    }
    #else
    for i in 0 ..< x.count {
        let xi = x[i]
        if xi == 0 { continue }
        let base = (rowOffset + i) * nOut
        for j in 0 ..< nOut { y[j] += kernel[base + j] * xi }
    }
    #endif
}

@inline(__always) private func load(_ destination: inout [Float], from source: [Float]) {
    destination.withUnsafeMutableBufferPointer { destinationBuffer in
        source.withUnsafeBufferPointer { sourceBuffer in
            destinationBuffer.baseAddress!.update(from: sourceBuffer.baseAddress!, count: sourceBuffer.count)
        }
    }
}

public final class Cell {
    public static let hidden = HIDDEN
    public static let attentionMixtures = ATTENTION_MIXTURES
    public static let outputMixtures = OUTPUT_MIXTURES
    public static let alphabetSize = ALPHABET_SIZE
    public static let maxChars = MAX_CHARS

    private let k1: [Float]
    private let b1: [Float]
    private let k2: [Float]
    private let b2: [Float]
    private let k3: [Float]
    private let b3: [Float]
    private let kAtt: [Float]
    private let bAtt: [Float]
    private let kGmm: [Float]
    private let bGmm: [Float]

    // Per-step scratch, reused across calls.
    private var gates = [Float](repeating: 0, count: GATES)
    private var attRaw = [Float](repeating: 0, count: 3 * ATTENTION_MIXTURES)
    private var gmmRaw = [Float](repeating: 0, count: 6 * OUTPUT_MIXTURES + 1)
    private var windowScratch = [Float](repeating: 0, count: ALPHABET_SIZE)
    private var inputScratch = [Float](repeating: 0, count: 3)
    private var matvecScratch = [Float](repeating: 0, count: GATES)

    public init(assets: ModelAssets) throws {
        func tensor(_ name: String) throws -> [Float] {
            guard let found = assets.tensors[name] else { throw CALWError.missingTensor(name) }
            return found.data
        }
        k1 = try tensor("lstm1_kernel")
        b1 = try tensor("lstm1_bias")
        k2 = try tensor("lstm2_kernel")
        b2 = try tensor("lstm2_bias")
        k3 = try tensor("lstm3_kernel")
        b3 = try tensor("lstm3_bias")
        kAtt = try tensor("attention_weights")
        bAtt = try tensor("attention_biases")
        kGmm = try tensor("gmm_weights")
        bGmm = try tensor("gmm_biases")
    }

    public func initialState() -> CellState {
        CellState()
    }

    public func newMdnParams() -> MdnParams {
        MdnParams()
    }

    /// In-place LSTM update from the fused gate buffer, TF gate order (i, j, f, o).
    private func applyLstm(hidden: inout [Float], cell: inout [Float]) {
        for m in 0 ..< HIDDEN {
            let inputGate = sigmoid(Double(gates[m]))
            let candidate = tanh(Double(gates[HIDDEN + m]))
            let forgetGate = sigmoid(Double(gates[2 * HIDDEN + m]))
            let outputGate = sigmoid(Double(gates[3 * HIDDEN + m]))
            let newCell = forgetGate * Double(cell[m]) + inputGate * candidate
            cell[m] = Float(newCell)
            hidden[m] = Float(outputGate * tanh(newCell))
        }
    }

    /// One timestep. Mutates `state` in place. `chars` holds alphabet indices
    /// (the encoded text, zero-padded to maxChars); `charLength` is the true
    /// encoded length.
    public func step(
        _ state: CellState,
        dx: Float,
        dy: Float,
        eos: Float,
        chars: [Int32],
        charLength: Int
    ) {
        inputScratch[0] = dx
        inputScratch[1] = dy
        inputScratch[2] = eos

        // LSTM 1: [w_prev, x, h1_prev]
        load(&gates, from: b1)
        accumulate(&gates, kernel: k1, rowOffset: 0, nOut: GATES, x: state.w, scratch: &matvecScratch)
        accumulate(&gates, kernel: k1, rowOffset: ALPHABET_SIZE, nOut: GATES, x: inputScratch, scratch: &matvecScratch)
        accumulate(&gates, kernel: k1, rowOffset: ALPHABET_SIZE + 3, nOut: GATES, x: state.h1, scratch: &matvecScratch)
        applyLstm(hidden: &state.h1, cell: &state.c1)

        // Attention: [w_prev, x, h1] -> softplus -> (alpha, beta, kappa step)
        load(&attRaw, from: bAtt)
        accumulate(&attRaw, kernel: kAtt, rowOffset: 0, nOut: 30, x: state.w, scratch: &matvecScratch)
        accumulate(&attRaw, kernel: kAtt, rowOffset: ALPHABET_SIZE, nOut: 30, x: inputScratch, scratch: &matvecScratch)
        accumulate(&attRaw, kernel: kAtt, rowOffset: ALPHABET_SIZE + 3, nOut: 30, x: state.h1, scratch: &matvecScratch)
        for i in 0 ..< 30 { attRaw[i] = Float(softplus(Double(attRaw[i]))) }
        for k in 0 ..< ATTENTION_MIXTURES {
            state.kappa[k] = Float(Double(state.kappa[k]) + Double(attRaw[20 + k]) / 25.0)
            if attRaw[10 + k] < 0.01 { attRaw[10 + k] = 0.01 } // beta floor
        }
        for u in 0 ..< MAX_CHARS {
            var sum = 0.0
            for k in 0 ..< ATTENTION_MIXTURES {
                let diff = Double(state.kappa[k]) - Double(u)
                sum += Double(attRaw[k]) * exp(-(diff * diff) / Double(attRaw[10 + k]))
            }
            state.phi[u] = Float(sum)
        }
        for i in 0 ..< ALPHABET_SIZE { windowScratch[i] = 0 }
        for u in 0 ..< charLength {
            windowScratch[Int(chars[u])] += state.phi[u]
        }
        load(&state.w, from: windowScratch)

        // LSTM 2: [x, h1, w, h2_prev]
        load(&gates, from: b2)
        accumulate(&gates, kernel: k2, rowOffset: 0, nOut: GATES, x: inputScratch, scratch: &matvecScratch)
        accumulate(&gates, kernel: k2, rowOffset: 3, nOut: GATES, x: state.h1, scratch: &matvecScratch)
        accumulate(&gates, kernel: k2, rowOffset: 3 + HIDDEN, nOut: GATES, x: state.w, scratch: &matvecScratch)
        accumulate(&gates, kernel: k2, rowOffset: 3 + HIDDEN + ALPHABET_SIZE, nOut: GATES, x: state.h2, scratch: &matvecScratch)
        applyLstm(hidden: &state.h2, cell: &state.c2)

        // LSTM 3: [x, h2, w, h3_prev]
        load(&gates, from: b3)
        accumulate(&gates, kernel: k3, rowOffset: 0, nOut: GATES, x: inputScratch, scratch: &matvecScratch)
        accumulate(&gates, kernel: k3, rowOffset: 3, nOut: GATES, x: state.h2, scratch: &matvecScratch)
        accumulate(&gates, kernel: k3, rowOffset: 3 + HIDDEN, nOut: GATES, x: state.w, scratch: &matvecScratch)
        accumulate(&gates, kernel: k3, rowOffset: 3 + HIDDEN + ALPHABET_SIZE, nOut: GATES, x: state.h3, scratch: &matvecScratch)
        applyLstm(hidden: &state.h3, cell: &state.c3)
    }

    /// MDN head with the Graves bias (sharpness) trick. Matches the reference:
    /// pi logits scaled by (1 + bias), log-sigmas shifted down by bias, then
    /// pi and eos snapped to zero below 0.01 (pi is NOT renormalized).
    public func mdnParse(h3: [Float], bias: Double, into out: MdnParams) {
        load(&gmmRaw, from: bGmm)
        accumulate(&gmmRaw, kernel: kGmm, rowOffset: 0, nOut: 6 * OUTPUT_MIXTURES + 1, x: h3, scratch: &matvecScratch)
        let raw = gmmRaw

        let M = OUTPUT_MIXTURES
        var maxLogit = -Double.infinity
        for m in 0 ..< M {
            let logit = Double(raw[m]) * (1 + bias)
            out.pi[m] = Float(logit)
            if logit > maxLogit { maxLogit = logit }
        }
        var total = 0.0
        for m in 0 ..< M {
            let value = exp(Double(out.pi[m]) - maxLogit)
            out.pi[m] = Float(value)
            total += value
        }
        for m in 0 ..< M {
            let p = Double(out.pi[m]) / total
            out.pi[m] = p < 0.01 ? 0 : Float(p)
            out.sigmaX[m] = Float(max(exp(Double(raw[M + m]) - bias), SIGMA_FLOOR))
            out.sigmaY[m] = Float(max(exp(Double(raw[2 * M + m]) - bias), SIGMA_FLOOR))
            let tanhRho = tanh(Double(raw[3 * M + m]))
            out.rho[m] = Float(min(max(tanhRho, EPSILON - 1), 1 - EPSILON))
            out.muX[m] = raw[4 * M + m]
            out.muY[m] = raw[5 * M + m]
        }
        let eosProbability = min(max(sigmoid(Double(raw[6 * M])), EPSILON), 1 - EPSILON)
        out.eos = eosProbability < 0.01 ? 0 : Float(eosProbability)
    }

    /// Draw one (Δx, Δy, eos) from parsed MDN params.
    public func mdnSample(_ params: MdnParams, rng: inout Rng) -> StrokeOffset {
        let component = rng.categorical(params.pi)
        let z1 = rng.normal()
        let z2 = rng.normal()
        let rho = Double(params.rho[component])
        let dx = Double(params.muX[component]) + Double(params.sigmaX[component]) * z1
        let dy = Double(params.muY[component])
            + Double(params.sigmaY[component]) * (rho * z1 + max(1 - rho * rho, 0).squareRoot() * z2)
        let eos = rng.uniform() < Double(params.eos)
        return StrokeOffset(dx: dx, dy: dy, eos: eos)
    }
}
